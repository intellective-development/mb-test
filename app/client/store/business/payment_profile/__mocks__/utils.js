// @flow

export const tokenizeCard = jest.fn();
