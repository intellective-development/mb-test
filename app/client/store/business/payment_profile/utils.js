// @flow

import braintree_client from 'braintree-web/client';

type CardData = {
  number: string,
  expirationMonth: string,
  expirationYear: string,
  cvv: string,
  billingAddress: {
    postalCode: string
  }
}
type TokenizeOptions = {
  braintree_client_token: string
}

const BRAINTREE_ERROR_BODY = {message: 'Invalid Credit Card'};
export const tokenizeCard = (card_data: CardData, {braintree_client_token}: TokenizeOptions): Promise<string> => {
  const tokenize_card_prom = new Promise((resolve, reject) => {
    braintree_client.create(
      { authorization: braintree_client_token },
      (create_error, client_instance) => {
        if (create_error){
          if (global.Raven) global.Raven.captureMessage(create_error);
          reject({error: BRAINTREE_ERROR_BODY});
          return null;
        }
        return client_instance.request({
          endpoint: 'payment_methods/credit_cards',
          method: 'post',
          data: { creditCard: card_data, options: { validate: false } }
        }, (request_error, response) => {
          if (request_error){
            if (global.Raven) global.Raven.captureMessage(request_error);
            reject({error: BRAINTREE_ERROR_BODY});
          } else {
            const nonce = response.creditCards[0].nonce;
            resolve(nonce);
          }
        });
      }
    );
  });

  return tokenize_card_prom;
};
