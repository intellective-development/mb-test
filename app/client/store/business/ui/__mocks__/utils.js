// @flow

export const enterStore = jest.fn();
