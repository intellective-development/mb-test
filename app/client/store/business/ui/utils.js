// @flow

export const enterStore = (initial_destination: string = '/store') => {
  let formatted_destination = initial_destination;
  if (formatted_destination.substr(0, 1) !== '/'){ // append to root
    formatted_destination = `/${formatted_destination}`;
  }

  const prefix = '/store'; // formatted_destination will now always have a leading slash, don't need trailing here
  // if formatted_destination doesn't start with /store, prepend it. for promos
  if (formatted_destination.substr(0, prefix.length) !== prefix){
    formatted_destination = prefix + formatted_destination;
  }

  window.location.href = formatted_destination.toLowerCase();
};
