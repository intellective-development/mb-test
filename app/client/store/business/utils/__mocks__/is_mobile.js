export const isIOS = jest.fn();
export const isAndroid = jest.fn();
export const isMobile = jest.fn();

export const __clearMocks__ = () => {
  isIOS.mockClear();
  isAndroid.mockClear();
  isMobile.mockClear();
};
