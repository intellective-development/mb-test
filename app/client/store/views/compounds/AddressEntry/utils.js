// @flow

import uuid from 'uuid';
import _ from 'lodash';
import type { Address } from 'store/business/address';

// This is a collection of pure functions, essentially abstracted out of the address entry module
// The goal is to have each of these have simple, single responsibility

type SubstringMatch = {length: number, offset: number};
type GoogleAddressTerm = {offset: number, value: string};

// NOTE this is a the type of objects returned from google.maps.places.AutocompleteService
// an is NOT the same as the GooglePlaceAddressComponent type defined below, which defines
// the structure of objects returned from google.maps.places.PlacesService
export type GoogleAddress = {
  description: string,
  id: string,
  matched_substrings: Array<SubstringMatch>,
  place_id: string,
  reference: string,
  structured_formatting: Object, // has main text and substring matches, same for secondary text
  terms: Array<GoogleAddressTerm>,
  types: Array<string> // more of an enum along the lines of 'street_address' | 'geocode' | ...
};

// TODO: should just use legacy_store address models displayShortAddress function
export const addressToString = (address?: Address | GoogleAddress): string => {
  if (!address) return '';
  if (address.description) return address.description; // if google result just return formatted string
  const {address1, address2, city, state, zip_code} = address;
  if (!address1 || !city || !state || !zip_code) return ''; // empty string if not all components present
  let main_text = address1;
  if (address2) main_text = `${main_text}, ${address2}`; // e.g. {mt: '560 Broadway, apt 2', st: 'New York NY 10002'}
  return `${main_text}, ${city} ${state} ${zip_code}`; // e.g. 560 Broadway, New York NY 10002
};

// TODO: write generalized regex special char stripping function
export const matchChunksInString = (full_string: string, chunks: Array<string>) => {
  let test_string = full_string;
  let acc_offset = 0;
  return chunks.filter(c => c.length > 0).reduce((acc, chunk) => { // remove empty strings
    const maybe_match = test_string.match(new RegExp(chunk, 'i'));
    if (!maybe_match) return acc;
    const offset = maybe_match.index;
    const length = maybe_match[0].length;
    const match = {offset: acc_offset + offset, length};
    test_string = test_string.slice(offset + length);
    acc_offset = acc_offset + offset + length;
    return [...acc, match];
  }, []);
};

export const addressToOptionTextProps = ({address1, address2, city, state, zip_code}: Address, input_string: string) => {
  if (!address1 || !city || !state || !zip_code) return ''; // empty string if not all components present
  let main_text = address1;
  if (address2) main_text = `${main_text}, ${address2}`; // e.g. {mt: '560 Broadway, apt 2', st: 'New York NY 10002'}
  const secondary_text = `${city} ${state} ${zip_code}`;
  const input_chunks = input_string ? input_string.split(' ') : [];
  return {
    main_text,
    secondary_text,
    main_text_matches: matchChunksInString(`${main_text},`, input_chunks),
    secondary_text_matches: matchChunksInString(secondary_text, input_chunks)
  };
};

// TODO: remove this in favor of Server deduping?
// just sticking the address1 and zip together so that both are being taken into account
export const uniqAddresses = (addresses: Array<Address>) => (
  _.uniqBy(addresses, (addr) => `${addr.address1}_${addr.zip_code}`)
);

// This is a collection of pure functions, essentially abstracted out of the address entry module
// The goal is to have each of these have simple, single responsibility

const component_names: Array<string> = [
  'street_number',
  'premise',
  'route',
  'sublocality',
  'locality',
  'administrative_area_level_1',
  'postal_code'
];

// type of the object that appears within the address_components array in a google.maps.places.PlacesService response
type GooglePlacesAddressComponent = {long_name: string, short_name: string, types: Array<string>};

export const getStoreableAddress = (response, addresses) => {
  const address_components = response && getAddressComponents(response.address_components);
  const formattedAddress = formatAddressForAPI(address_components, response.geometry);
  return findStoreableAddress(formattedAddress, addresses);
};

export const findStoreableAddress = (formattedAddress, addresses) => {
  const savedAddresses = _.filter(_.values(addresses), ({ id }) => !!id);
  const foundAddresses = _.filter(savedAddresses, _.omit(formattedAddress, ['local_id', 'latitude', 'longitude']));
  // sort those that have id up (descending), pick first, if not fallback to creating new
  return _.head(_.orderBy(foundAddresses, ['id'], ['desc'])) || {local_id: uuid(), ...formattedAddress};
};

// convert googles format to object keyed by the types we want
export const getAddressComponents = (address_components: Array<GooglePlacesAddressComponent>): GoogleAddressObject => {
  const address_obj = {};

  // create an object from the array (based on the type subarray)
  _.forEach(component_names, component_name => {
    address_obj[component_name] = _.find(address_components, component => (
      _.includes(component.types, component_name)
    ));
  });

  address_obj.street_number = address_obj.street_number || address_obj.premise;
  address_obj.city = address_obj.sublocality || address_obj.locality;

  return address_obj;
};

// our object constructed from the parsed google.maps.places.PlacesService result
type GoogleAddressObject = {
  street_number?: GooglePlacesAddressComponent,
  premise?: GooglePlacesAddressComponent,
  route?: GooglePlacesAddressComponent,
  sublocality?: GooglePlacesAddressComponent,
  locality?: GooglePlacesAddressComponent,
  administrative_area_level_1?: GooglePlacesAddressComponent,
  postal_code?: GooglePlacesAddressComponent,
  city?: GooglePlacesAddressComponent
}
type GooglePlacesGeometry = {
  location: {lat: () => string, lng: () => string},
  viewport?: {
    northeast?: {lat: () => string, lng: () => string},
    southwest?: {lat: () => string, lng: () => string}
  }
}

// map to false if no street #
export const validateComponents = (address: GoogleAddressObject) => (address.street_number ? address : false);

export const formatAddressForAPI = (address: GoogleAddressObject, geometry: GooglePlacesGeometry) => {
  const street_number = formatAddressComponent(address.street_number);
  const road = formatAddressComponent(address.route);
  const city = formatAddressComponent(address.city);
  const state = formatAddressComponent(address.administrative_area_level_1, 'short_name');
  const zip_code = formatAddressComponent(address.postal_code);

  // TODO:  Test state formatter
  return {
    address1: `${street_number} ${road}`,
    city: city,
    state: state,
    zip_code: zip_code,
    latitude: geometry.location.lat(),
    longitude: geometry.location.lng()
  };
};

export const formatAddressComponent = (component?: GooglePlacesAddressComponent, length?: string = 'long_name') => (
  component ? component[length] : ''
);

// This creates a memoized function that takes a string1 and string2 and returns true
// if string2's characters appear in order within string1 with 1 char of fuzziness at the end
export const makeFuzzyMatcher = () => {
  const matcherCache = _.memoize(str => ( // interpolate every character with '?' (0 or more matcher)
    new RegExp(`^${str.replace(/./gi, x => (
      /[-[\]/{}()*+?.\\^$|]/i.test(x) ? `\\${x}?` : `${x}?` // escape special characters
    ))}.{1}$`, 'i') // .{1} gives one char of terminal fuzziness, 'i' is case insensitive
  )); // e.g. input '45 Bond' => output new RegExp('^4?5? ?B?o?n?d?.{1}$', 'i')
  return (string1: string, string2: string) => matcherCache(string1).test(string2);
};

export const nonCurrentAddresses = (addresses?: Array<Address>, current_address?: Address): Array<Address> => (
  addresses ? addresses.filter(a => a.local_id !== _.get(current_address, 'local_id')) : []
);
