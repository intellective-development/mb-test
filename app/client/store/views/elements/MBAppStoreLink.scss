
// we apply a minimal layout here, expecting the parent to decide how they should be laid out
.elMBAppStoreLink_Container {
  display: flex;
  justify-content: space-between;
}
