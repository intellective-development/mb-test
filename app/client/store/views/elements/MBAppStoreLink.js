// @flow

import * as React from 'react';
import bindClassNames from '../../../shared/utils/bind_classnames';
import { isIOS, isAndroid } from '../../business/utils/is_mobile';

import MBLink from './MBLink';
import styles from './MBAppStoreLink.scss';

const cn = bindClassNames(styles);
const IOS_APP_STORE_LINK = 'https://itunes.apple.com/us/app/minibar-delivery-wine-liquor/id720850888?mt=8&uo=4&at=11l5XB&ct=minibar';
const PLAY_STORE_LINK = 'https://play.google.com/store/apps/details?id=minibar.android';

type MBAppStoreLinkProps = {className?: string};
const MBAppStoreLink = ({className}: MBAppStoreLinkProps) => {
  if (isIOS()) return <div className={cn('elMBAppStoreLink_Container', className)}><MBiOSAppStoreLink /></div>;
  if (isAndroid()) return <div className={cn('elMBAppStoreLink_Container', className)}><MBPlayStoreLink /></div>;

  // TODO: return fragment
  return (
    <div className={cn('elMBAppStoreLink_Container', className)}>
      <MBiOSAppStoreLink />
      <MBPlayStoreLink />
    </div>
  );
};

export const APP_STORE_LINK_ID = 'app-store-link';
const MBiOSAppStoreLink = () => (
  <MBLink.View id={APP_STORE_LINK_ID} href={IOS_APP_STORE_LINK} native_behavior target="_blank" rel="noopener noreferrer">
    <img
      alt="Download on the App Store"
      src={'/assets/components/elements/mb-app-store-link/ios-app-store.png'}
      srcSet={'/assets/components/elements/mb-app-store-link/ios-app-store@2x.png 2x, ' +
              '/assets/components/elements/mb-app-store-link/ios-app-store@3x.png 3x'} />
  </MBLink.View>
);

export const PLAY_STORE_LINK_ID = 'play-store-link';
const MBPlayStoreLink = () => (
  <MBLink.View id={PLAY_STORE_LINK_ID} href={PLAY_STORE_LINK} native_behavior target="_blank" rel="noopener noreferrer">
    <img
      alt="Get it on Google Play"
      src={'/assets/components/elements/mb-app-store-link/play-store.png'}
      srcSet={'/assets/components/elements/mb-app-store-link/play-store@2x.png 2x, ' +
              '/assets/components/elements/mb-app-store-link/play-store@3x.png 3x'} />
  </MBLink.View>
);

export default MBAppStoreLink;
