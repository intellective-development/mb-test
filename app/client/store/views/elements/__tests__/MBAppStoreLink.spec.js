
import React from 'react';

import { isIOS, isAndroid, __clearMocks__ } from 'store/business/utils/is_mobile';
import MBAppStoreLink from '../MBAppStoreLink';

jest.mock('store/business/utils/is_mobile');

describe('MBAppStoreLink', () => {
  afterEach(() => {
    __clearMocks__();
  });

  it('renders an app store button if the userAgent matches ios', () => {
    isIOS.mockReturnValue(true);
    isAndroid.mockReturnValue(false);

    expect(render(
      <MBAppStoreLink />
    )).toMatchSnapshot();
  });

  it('renders a google play store button if the userAgent matches android', () => {
    isIOS.mockReturnValue(false);
    isAndroid.mockReturnValue(true);

    expect(render(
      <MBAppStoreLink />
    )).toMatchSnapshot();
  });

  it('renders both if platform can not be determined by userAgent', () => {
    isIOS.mockReturnValue(false);
    isAndroid.mockReturnValue(false);

    expect(render(
      <MBAppStoreLink />
    )).toMatchSnapshot();
  });
});
