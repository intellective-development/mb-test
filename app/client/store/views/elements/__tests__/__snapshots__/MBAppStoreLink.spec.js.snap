// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MBAppStoreLink renders a google play store button if the userAgent matches android 1`] = `
<div
  class="elMBAppStoreLink_Container"
>
  <a
    href="https://play.google.com/store/apps/details?id=minibar.android"
    id="play-store-link"
    rel="noopener noreferrer"
    target="_blank"
  >
    <img
      alt="Get it on Google Play"
      src="/assets/components/elements/mb-app-store-link/play-store.png"
      srcset="/assets/components/elements/mb-app-store-link/play-store@2x.png 2x, /assets/components/elements/mb-app-store-link/play-store@3x.png 3x"
    />
  </a>
</div>
`;

exports[`MBAppStoreLink renders an app store button if the userAgent matches ios 1`] = `
<div
  class="elMBAppStoreLink_Container"
>
  <a
    href="https://itunes.apple.com/us/app/minibar-delivery-wine-liquor/id720850888?mt=8&uo=4&at=11l5XB&ct=minibar"
    id="app-store-link"
    rel="noopener noreferrer"
    target="_blank"
  >
    <img
      alt="Download on the App Store"
      src="/assets/components/elements/mb-app-store-link/ios-app-store.png"
      srcset="/assets/components/elements/mb-app-store-link/ios-app-store@2x.png 2x, /assets/components/elements/mb-app-store-link/ios-app-store@3x.png 3x"
    />
  </a>
</div>
`;

exports[`MBAppStoreLink renders both if platform can not be determined by userAgent 1`] = `
<div
  class="elMBAppStoreLink_Container"
>
  <a
    href="https://itunes.apple.com/us/app/minibar-delivery-wine-liquor/id720850888?mt=8&uo=4&at=11l5XB&ct=minibar"
    id="app-store-link"
    rel="noopener noreferrer"
    target="_blank"
  >
    <img
      alt="Download on the App Store"
      src="/assets/components/elements/mb-app-store-link/ios-app-store.png"
      srcset="/assets/components/elements/mb-app-store-link/ios-app-store@2x.png 2x, /assets/components/elements/mb-app-store-link/ios-app-store@3x.png 3x"
    />
  </a>
  <a
    href="https://play.google.com/store/apps/details?id=minibar.android"
    id="play-store-link"
    rel="noopener noreferrer"
    target="_blank"
  >
    <img
      alt="Get it on Google Play"
      src="/assets/components/elements/mb-app-store-link/play-store.png"
      srcset="/assets/components/elements/mb-app-store-link/play-store@2x.png 2x, /assets/components/elements/mb-app-store-link/play-store@3x.png 3x"
    />
  </a>
</div>
`;
